Refactor `pytest_addoption`, expose `rp_uuid` and `rp_endpoint` to command line

rthardin opened the pull request:
This merge request aims to allow the `rp_uuid` and `rp_endpoint` to be set using the command line. This MR consists of three commits:

1. Add unit tests which verify which options are added to the `.ini` file and the command line.
2. Refactor `pytest_addoption` to make it easier to keep both sets of arguments consistent.
3. Add `rp_uuid` and `rp_endpoint` as command line arguments.

rthardin commented:
It looks like I will also need to change config.py to pick up the changes to rp_uuid and rp_endpoint. I can post an update to the most recent commit for that on Tuesday.

iivanou commented on tests/test_plugin.py:
```diff
@@ -230,3 +231,71 @@ def test_pytest_unconfigure(mocked_config):
+
+def test_pytest_addoption_adds_correct_ini_file_arguments():
+    """Test the correct list of options are available in the .ini file."""
+    expected_argument_names = [
```
May be converted to a tuple. It consumes less memory.

rthardin replied:
Done ✨ Thanks!

iivanou commented on tests/test_plugin.py:
```diff
@@ -230,3 +231,71 @@ def test_pytest_unconfigure(mocked_config):
+        'rp_verify_ssl',
+        'rp_display_suite_test_file',
+        'rp_issue_id_marks',
+        'retries',
```
A comma can be omitted.

rthardin replied:
Done ✨ Thanks!

iivanou commented on tests/test_plugin.py:
```diff
@@ -1,22 +1,23 @@
     pytest_sessionfinish,
     pytest_unconfigure,
-    wait_launch
+    wait_launch,
```
A comma can be omitted.

rthardin replied:
Done ✨ Thanks!

iivanou commented on tests/test_plugin.py:
```diff
@@ -1,22 +1,23 @@
-
-from delayed_assert import expect, assert_expectations
 import pytest
+from _pytest.config.argparsing import Parser
```
Put it before `import pytest`, please.

rthardin replied:
Done ✨ Thanks!

iivanou commented on tests/test_plugin.py:
```diff
@@ -1,22 +1,23 @@
 import pytest
+from _pytest.config.argparsing import Parser
+from delayed_assert import expect, assert_expectations
+from reportportal_client.errors import ResponseError
```
https://www.python.org/dev/peps/pep-0008/#imports

Imports should be grouped in the following order:

- Standard library imports.
- Related third party imports.
- Local application/library specific imports.

You should put a blank line between each group of imports.

rthardin replied:
Are these not all third party imports? I can split them however you'd like, but I don't think any of them are part of the standard library.

iivanou replied:
reportportal_client is a local application.

rthardin replied:
I still don't understand why that's the case since `reportportal_client` comes from `requirements.txt`, but I have reverted the line.

iivanou replied:
Correct, that's my bad. I thought it is a common client where we are))

iivanou reviewed (changes requested):
Looks good in general. Just a few small corrections are required.

rthardin commented:
Thanks for your feedback, @iivanou . I have updated each of those items, and have left them unresolved for you to check over. I have updated the second commit to include the changes to `config.py` which adds a `find_option` function. Hopefully this will make it easier to maintain options which live in both the command line and the `.ini` file. While I was at it, I updated some of the unit tests by moving some mock code into the fixture and adding a few extra checks.

iivanou commented on pytest_reportportal/config.pyi:
```diff
@@ -23,3 +23,5 @@ class AgentConfig:
     @property
     def rp_rerun(self) -> Optional[bool]: ...
+
+    def find_option(self, param: str): ...
```
```suggestion
    def find_option(self, param: str) -> Union[str, bool]: ...
```

rthardin replied:
Done ✨ Thanks!

iivanou commented:
@rthardin Can you please update README accordingly? Thank you.
